util/network/src/client_version.rs

```diff
@@ -0,0 +1,473 @@
+	let service_transactions_version = Version::parse(SERVICE_TRANSACTIONS_VERSION).unwrap();
+
+	*parity_client_data.semver() >= service_transactions_version
+}
+
+impl ClientCapabilities for ClientVersion {
+	fn can_handle_large_requests(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => {
+				if *data.semver() < Version::parse(PARITY_CLIENT_LARGE_REQUESTS_VERSION).unwrap() {
+					false
+				} else {
+					true
+				}
+			},
+			ClientVersion::ParityUnknownFormat(_) => false, // Play it safe
+			ClientVersion::Other(_) => true // As far as we know
+		}
+	}
+
+	/// Checks if peer is able to process service transactions
+	fn accepts_service_transaction(&self) -> bool {
+		match self {
+			ClientVersion::ParityClient(data) => parity_accepts_service_transaction(&data),
+			ClientVersion::ParityUnknownFormat(_) => false,
+			ClientVersion::Other(_) => false
+		}
+	}
+
+}
+
+
+fn is_parity(client_id: &str) -> bool {
+	if client_id.starts_with(LEGACY_CLIENT_ID_PREFIX) || client_id.starts_with(PARITY_CLIENT_ID_PREFIX) {
+		return true;
+	} else {
+		return false;
+	}
+}
+
+// Parse known parity formats.
+//
+// This is really not robust: parse four arguments and
+// allow for an extra argument between identifier and
+// version
+// TODO implement a better logic
+fn parse_parity_format(client_version: &str) -> Result<ParityClientData, ()> {
+	let tokens: Vec<&str> = client_version.split("/").collect();
+
+
+	// Basically strip leading 'v'
+	if let Some(version_number) = &get_number_from_version(tokens[1]) {
+		if tokens.len() < 4 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: None,
+				semver: Version::parse(version_number).unwrap(),
+				os: tokens[2].to_string(),
+				compiler: tokens[3].to_string(),
+			}
+		);
+	} else if let Some(version_number) = &get_number_from_version(tokens[2]) {
+		if tokens.len() < 5 {
+			return Err(());
+		}
+
+		return Ok(
+			ParityClientData {
+				name: tokens[0].to_string(),
+				variant: Some(tokens[1].to_string()),
+				semver: Version::parse(version_number).unwrap(),
+				os: tokens[3].to_string(),
+				compiler: tokens[4].to_string(),
+			}
+		);
+	} else {
+		return Err(());
+	}
+}
+
+// Parses a version string and returns the corresponding
+// ClientVersion. Only Parity clients are destructured right now.
+// The parsing for parity may still fail, in which case return an Other with
+// the original version string. TryFrom would be a better trait to implement.
+
+impl From<&str> for ClientVersion {
+	fn from(client_version: &str) -> Self {
+		if !is_parity(client_version) {
+			return ClientVersion::Other(client_version.to_string());
+		}
+
+		if let Ok(data) = parse_parity_format(client_version) {
+			ClientVersion::ParityClient(data)
+		} else {
+			ClientVersion::ParityUnknownFormat(client_version.to_string())
+		}
+	}
+}
+
+impl From<String> for ClientVersion {
```

Comment:
Might be unnecessary if we implement for `T: AsRef<str>` I think

Reply:
Ok, I'll look into that.

Reply:
What do you mean? `impl From<T: AsRef<str>> for CV`? Or you mean implementing AsRef<str> for CV? I can see the first alternative being more general than this implementation, and the second one would be an alternative for `fmt::Display` I guess?

Reply:
I meant the former, to just have one `From` implementation (instead of `From<String>` and `From<&str>` as we have now), but it's nitpicking :)

Reply:
Oh, I see. I'm relatively new to rust, so I'm glad to learn the tricks ;-)

Reply:
Yep, it did the trick. Nice :-)